c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -482,19 +484,131 @@ class CachingAllocatorConfig {
+class MemoryPlanner {
+ private:
+  std::vector<std::vector<AllocFreeEvent>>
+      mem_plan; // sequence of expected allocate events for each device
+  std::vector<Block*>
+      plan_blk_ptr; // ptr for block used to execute the plan for each device
+  std::vector<unsigned int>
+      plan_counter; // points to the next expected allocation
+  std::vector<std::map<intptr_t, bool>>
+      allocated_blocks; // stores start and end of each allocated block to avoid
+                        // overlaps
+  std::vector<size_t>
+      planning_size; // size of block used to execute the plan for each device
+  std::vector<uint32_t>
+      non_rec_length; // number of transient events in the beginning of the plan
+                      // that are not expected to be periodic
+  std::mutex mutex;
+
+ public:
+  std::map<void*, Block*>
+      assigned_planned_blocks; // stores allocations done by the plan to call
+                               // the approperiate free
+  bool active = false; // if there is plan for at least one device
+  void create_plan(
+      const std::vector<std::vector<AllocFreeEvent>>& plan,
+      std::vector<Block*> plan_ptrs,
+      std::vector<size_t> plan_size,
+      std::vector<uint32_t> non_rec_lengths) {
+    std::lock_guard<std::mutex> lock(mutex);
+    mem_plan = plan;
+    plan_blk_ptr.resize(plan.size(), nullptr);
+    plan_counter.resize(plan.size(), 1);
+    planning_size.resize(plan.size(), 0);
+    allocated_blocks.resize(plan.size());
+    non_rec_length.resize(plan.size(), 0);
+    plan_blk_ptr = plan_ptrs;
+    planning_size = plan_size;
+    non_rec_length = non_rec_lengths;
+    active = true;
+  }
+
+  Block* malloc(size_t size, int device, cudaStream_t stream) {
+    Block* block =
+        nullptr; // if block remains null, this indicates event not allocated by
+                 // plane, hence, use CudaCachingAllocator
+    // event is allocated according to plan if its size matches the expected
+    // allocation size
+    if ((plan_blk_ptr[device] != nullptr) &&
+        (size == (uint32_t)mem_plan[device][plan_counter[device]].size)) {
+      intptr_t planned_ptr = mem_plan[device][plan_counter[device]]
+                                 .ptr; // allocation offset according to plan
+      auto it_ptr_lower_bound = allocated_blocks[device].lower_bound(
+          planned_ptr); // find the closest start/end point of a previous
+                        // allocation after the current allocate ptr
+      // perform the allocation only if it fails inside the block assigned for
+      // plan execution and it does not overlap with previuosly allocated events
+      // if it_ptr_lower_bound points to an end point, this means the current
+      // allocation starts in the middle of a previous allocation; do not
+      // allocate. If it_ptr_lower_bound points to a start point, check if
+      // planned_ptr + size <= start point to ensure no overlap
+      if ((planned_ptr + size <= planning_size[device]) &&
```

Comment:
This is just to be extra careful that the memory plan is feasible since it is passed from the python side and we do not have control over it. If some allocations are not feasible according to plan, we pass those allocations to the normal CudaCachingAllocator.